llvm/docs/Mdl/MachineDescriptionNotes.md

```diff
@@ -0,0 +1,2483 @@
+
+Grammar:
+
+```
+    derived_operand_def     : 'operand' IDENT (':' IDENT)+  ('(' ')')?
+                                  '{' (operand_type | operand_attribute)* '}' ';'? ;
+    operand_attribute_stmt  : 'attribute' IDENT '=' (snumber | tuple)
+                                ('if' ('lit' | 'address' | 'label')
+                                ('[' pred_value (',' pred_value)* ']' )? )? ';' `;
+
+    pred_value              : snumber
+                            | snumber '..' snumber
+                            | '{' number '}' ;
+    tuple                   : '[' snumber (',' snumber)* ']' ;
+```
+
+#### **Derivation**
+
+Each derived operand is declared with one or more “base” operands, for which it is an alias. Circular or ambiguous derivations are explicitly disallowed - there must be only one derivation path for a derived operand to any of its base concrete operands.
+
+Derived operands are used in place of their base operands in operand latency rules in latency templates (described later). This allows a rule to match a set of operands, rather than a single operand, and also can provide access to instruction attributes to the latency rule.
+
+#### **Derived operand attributes**
+
+Derived operand attributes associate name/value-tuple pairs with the operand type. Tuples are appropriate when an attribute is used as a set of masks for resource sharing, described later.  
+
+Some examples:
+
+```
+    attribute my_attr_a = 1;
+    attribute my_attr_b = 123;
+    attribute my_tuple  = [1, 2, 3];
+```
+
+Attributes can have predicates that check if the operand contains a data address, a code address, or any constant.  Additionally, attributes can have multiple definitions with different predicates, with the first “true” predicate determining the final value of the attribute for that operand instance:
+
+```
+    attribute my_attr = 5 if address;    // if operand is a relocatable address
+    attribute my_attr = 2 if label;      // if operand is a code address
+    attribute my_attr = 3 if lit;        // if operand is any literal constant
+```
+
+
+Predicates for literal constants can also take an optional list of “predicate values”, where each predicate value is either an integer, a range of integers, or a “mask”. Mask predicate values are explicitly checking for non-zero bits:
+
+```
+    attribute my_attr = 5 if lit [1, 2, 4, 8];    // looking for specific values
+    attribute my_attr = 12 if lit [100..200];     // looking for a range of values
+    attribute my_attr = 1 if lit [{0x0000FFFF}];  // looking for a 16 bit number
+    attribute my_attr = 2 if lit [{0x00FFFF00}];  // also a 16-bit number!
+    attribute my_attr = 3 if lit [1, 4, 10..14, 0x3F800000, {0xFF00FF00}]; 
```

Comment:
What if you need to describe something that can't be easily expressed with ranges like a 32-bit even number.

Reply:
You can use masks ( {....} ) for the "even" example: {0xFFFFFFFE} :-)
Clearly, this isn't fully general. The goal was to support just about anything I had ever seen.  Note that I also didn't want to mess directly with floating point, but you can enumerate the binaries for that.  So I picked the syntax to reflect the bits that go in the instruction.

Of course we can always make it more comprehensive.

Reply:
Oh I missed the mask part.

Reply:
FWIW, this is kind of an "esoteric" feature - its used to model the allocation of encoding resources related to any kind of "immediate" operand.  In VLIWs, this can get really messy when immediate bits are shared across the bundle.  So its not used for any upstream processor (but could be used for Hexagon, for example).

There are other ways to model latencies based on operand types and operand values (like your divide example).